### PR TITLE
QuestionAnsweringTask and QuestionAnsweringData with SQuADDatSource.

### LINKED ISSUE #540: Add a question_answering example

### LINKED ISSUE #541: Add a Question Answering page to the docs

### LINKED ISSUE #577: Add a SquadDataSource for QuestionAnsweringData and `from_squad` method
## 🚀 Feature

Add a data source for handling QA data in the squad format, similar to what we have for COCO in object detection.

### PR DESCRIPTION
**Pull request by karthikrangasai**
## What does this PR do?

Fixes #577
Fixes #540 
Fixes #541

- Update / rewrite `QuestionAnsweringTask` and `QuestionAnsweringData` to handle HF AutoModelForQuestionAnswering as backbone.
- Provide SQuAD Dataset as a `SQuADDataSource`.
- Add examples and documentation for both the Task and Data classes.
- Added tests (needs some work here).

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/text/seq2seq/core/data.py [hunk 1]
```diff
@@ -222,6 +223,62 @@ def __setstate__(self, state):
...
+        return model_inputs
+
+    def load_data(self, data: Any, columns: List[str] = None) -> 'datasets.Dataset':
+        if columns is None:
+            columns = ["input_ids", "attention_mask", "labels"]
+            if self._running_stage.value == RunningStage.PREDICTING:
+                columns.remove("labels")
+
+        stage = self._running_stage.value
+        data, input, input_pair, target = data
```
Comment (ethanwharris) [inline flash/text/seq2seq/core/data.py:262]: Let's just enforce `input="question"`, `input_pair="context"` and `target=None` (i.e. this will only be used for prediction). That way users can just pass the dictionary to the predict function rather than the tuple.

File: flash/text/seq2seq/question_answering/data.py [hunk 2]
```diff
@@ -29,19 +94,143 @@ def __init__(
...
+                    max_source_length=max_source_length,
+                    max_target_length=max_target_length,
+                    padding=padding,
+                ),
+                "sentences": Seq2SeqSentencesDataSource(
+                    self.backbone,
+                    max_source_length=max_source_length,
+                    max_target_length=max_target_length,
+                    padding=padding,
+                ),
```
Comment (ethanwharris) [inline flash/text/seq2seq/question_answering/data.py:125]: Should we remove these? Do they work for QA? Maybe just remove the sentences one if the CSV and JSON would still work.

File: flash/text/seq2seq/question_answering/data.py [hunk 3]
```diff
@@ -29,19 +94,143 @@ def __init__(
...
+            test_ann_file: The COCO format annotation file.
+            train_transform: The dictionary of transforms to use during training which maps
+                :class:`~flash.core.data.process.Preprocess` hook names to callable transforms.
+            val_transform: The dictionary of transforms to use during validation which maps
+                :class:`~flash.core.data.process.Preprocess` hook names to callable transforms.
+            test_transform: The dictionary of transforms to use during testing which maps
+                :class:`~flash.core.data.process.Preprocess` hook names to callable transforms.
+            data_fetcher: The :class:`~flash.core.data.callback.BaseDataFetcher` to pass to the
+                :class:`~flash.core.data.data_module.DataModule`.
+            preprocess: The :class:`~flash.core.data.data.Preprocess` to pass to the
```
Comment (ethanwharris) [inline flash/text/seq2seq/question_answering/data.py:203]: Looks like this docstring needs updating

**Review by ethanwharris — COMMENTED** [review]
Awesome work @karthikrangasai just a few small comments

File: flash/text/seq2seq/question_answering/data.py (hunk 2 above)
Comment (karthikrangasai) [reply flash/text/seq2seq/question_answering/data.py:125]: I think we can remove the sentences data source.

File: flash/text/question_answering/data.py [hunk 4]
```diff
@@ -0,0 +1,1017 @@
...
+        self._context_column_name = context_column_name
+        self._answer_column_name = answer_column_name
+        self._doc_stride = doc_stride
+
+    def _tokenize_fn(self, samples: Any) -> Callable:
+        stage = self._running_stage.value
+
+        samples[self.question_column_name] = [q.lstrip() for q in samples[self.question_column_name]]
+
+        self.pad_on_right = self.tokenizer.padding_side == "right"
```
Comment (ethanwharris) [inline flash/text/question_answering/data.py:179]: Best to define this in init

File: flash/text/question_answering/data.py [hunk 5]
```diff
@@ -0,0 +1,1017 @@
...
+            - context_column_name: The key in the JSON file to recognize the context field. Defaults to "context".
+            - answer_column_name: The key in the JSON file to recognize the answer field. Defaults to "answer".
+            - doc_stride: The stride amount to be taken when splitting up a long document into chunks.
+
+        Returns:
+            The constructed data module.
+
+        Examples::
+
+            data_module = DataModule.from_csv(
```
Comment (ethanwharris) [inline flash/text/question_answering/data.py:983]: These can all be `QuestionAnsweringData`
```suggestion
            data_module = QuestionAnsweringData.from_csv(
```

File: flash/text/question_answering/model.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                offsets = pred.pop("offsets")
+                pred["text"] = context[offsets[0] : offsets[1]]
+
+            # In the very rare edge case we have not a single non-null prediction, we create a fake prediction to avoid
+            # failure.
+            if len(predictions) == 0 or (len(predictions) == 1 and predictions[0]["text"] == ""):
+                predictions.insert(0, {"text": "empty", "start_logit": 0.0, "end_logit": 0.0, "score": 0.0})
+
+            # Compute the softmax of all scores (we do it with numpy to stay independent from torch/tf in this file,
+            # using the LogSumExp trick).
```
Comment (ethanwharris) [inline flash/text/question_answering/model.py:197]:
```suggestion
            # Compute the softmax of all scores.
```

File: flash/text/question_answering/model.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+            scores: Tensor = torch.tensor([pred.pop("score") for pred in predictions])
+            exp_scores: Tensor = torch.exp(scores - torch.max(scores))
+            probs: Tensor = exp_scores / exp_scores.sum()
```
Comment (ethanwharris) [inline flash/text/question_answering/model.py:200]: Can we just use torch.softmax here?

File: flash_examples/question_answering.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+model = QuestionAnsweringTask()
+
+# 3. Create the trainer and finetune the model
+trainer = Trainer(max_epochs=3, limit_train_batches=1, limit_val_batches=1)
+trainer.finetune(model, datamodule=datamodule)
+
+# 4. Answer some Questions!
+predictions = model.predict(
+    {
+        "id": ["56ddde6b9a695914005b9629", "56ddde6b9a695914005b9628"],
```
Comment (ethanwharris) [inline flash_examples/question_answering.py:37]: Is it possible to not provide the id when predicting? Ideally we would just provide the context and the question. Could also be nice to add support for context + list of questions, but that could be done in a follow-up PR

**Review by ethanwharris — APPROVED** [review]
Really awesome work!!!! Just a few small comments, maybe just remove the deserializer / serve stuff for now and we can add it in a follow-up PR with an example?

**Review by SeanNaren — APPROVED** [review]

**Comment by karthikrangasai** [comment]
Sure, I will remove the serve and deserializer code.

File: flash_examples/question_answering.py (hunk 8 above)
Comment (ethanwharris) [reply flash_examples/question_answering.py:37]: To be done in follow-up

Comment (karthikrangasai) [reply flash_examples/question_answering.py:37]: Currently the `id` is being used to reference the output accordingly in predict.
In a follow-up PR along with `context, list[questions]`  we can also provide non `id` based input.
